File: backend/app/services/inventory_service.py
from uuid import UUID
from typing import Optional
from decimal import Decimal

from sqlalchemy.orm import Session

from app.db.models.inventory_item import InventoryItem
from app.db.models.inventory_movement import InventoryMovement, MovementType
from app.repositories.inventory_repository import InventoryRepository


class InventoryService:
    def __init__(self, db: Session):
        self.repo = InventoryRepository(db)

    def get_item_by_id(self, item_id: UUID) -> Optional[InventoryItem]:
        return self.repo.get_item_by_id(item_id)

    def get_all_items(self, skip: int = 0, limit: int = 100) -> list[InventoryItem]:
        return self.repo.get_all_items(skip, limit)

    def get_low_stock(self) -> list[InventoryItem]:
        return self.repo.get_low_stock()

    def create_item(self, name: str, unit: str, quantity: Decimal = 0,
                    min_stock: Decimal = 0) -> InventoryItem:
        item = InventoryItem(name=name, unit=unit, quantity=quantity, min_stock=min_stock)
        return self.repo.create_item(item)

    def update_item(self, item_id: UUID, data: dict) -> Optional[InventoryItem]:
        item = self.repo.get_item_by_id(item_id)
        if not item:
            return None
        for key, value in data.items():
            if value is not None and hasattr(item, key):
                setattr(item, key, value)
        return self.repo.update_item(item)

    def register_movement(self, item_id: UUID, movement_type: str,
                        quantity: Decimal, reason: Optional[str] = None) -> Optional[InventoryMovement]:
        item = self.repo.get_item_by_id(item_id)
        if not item:
            return None
        movement = InventoryMovement(
            item_id=item_id, type=MovementType(movement_type),
            quantity=quantity, reason=reason,
        )
        if movement_type == "in":
            item.quantity += quantity
        else:
            item.quantity -= quantity
        self.repo.update_item(item)
        return self.repo.create_movement(movement)

    def get_movements(self, item_id: UUID) -> list[InventoryMovement]:
        return self.repo.get_movements_by_item(item_id)

File: frontend/src/services/inventoryService.js
import { apiGet, apiPost, apiPut, apiDelete } from "./api.js";

function mapItem(item) {
  return {
    id: item.id,
    name: item.name,
    unit: item.unit,
    quantity: parseFloat(item.quantity),
    min_stock: parseFloat(item.min_stock),
    is_active: item.is_active !== false,
    created_at: item.created_at,
    updated_at: item.updated_at,
  };
}

function mapMovement(m) {
  return {
    id: m.id,
    item_id: m.item_id,
    type: m.type,
    quantity: parseFloat(m.quantity),
    reason: m.reason || "",
    created_at: m.created_at,
  };
}

export async function getAllItems() {
  try {
    const items = await apiGet("/api/v1/inventory/");
    return items.map(mapItem);
  } catch {
    return [];
  }
}

export async function getItemById(id) {
  try {
    const item = await apiGet(`/api/v1/inventory/${id}`);
    return mapItem(item);
  } catch {
    return null;
  }
}

export async function getLowStockItems() {
  try {
    const items = await apiGet("/api/v1/inventory/low-stock");
    return items.map(mapItem);
  } catch {
    return [];
  }
}

export async function createItem(data) {
  try {
    const item = await apiPost("/api/v1/inventory/", {
      name: data.name,
      unit: data.unit,
      quantity: data.quantity || 0,
      min_stock: data.min_stock || 0,
    });
    return { success: true, item: mapItem(item) };
  } catch (err) {
    return { success: false, error: err.message };
  }
}

export async function updateItem(id, data) {
  try {
    const item = await apiPut(`/api/v1/inventory/${id}`, {
      name: data.name,
      unit: data.unit,
      quantity: data.quantity,
      min_stock: data.min_stock,
    });
    return { success: true, item: mapItem(item) };
  } catch (err) {
    return { success: false, error: err.message };
  }
}

export async function registerMovement(itemId, data) {
  try {
    const movement = await apiPost(`/api/v1/inventory/${itemId}/movements`, {
      type: data.type,
      quantity: data.quantity,
      reason: data.reason || "",
    });
    return { success: true, movement: mapMovement(movement) };
  } catch (err) {
    return { success: false, error: err.message };
  }
}

export async function deleteItem(id) {
  try {
    await apiDelete(`/api/v1/inventory/${id}`);
    return { success: true };
  } catch (err) {
    return { success: false, error: err.message };
  }
}

export async function getMovementsByItem(itemId) {
  try {
    const items = await apiGet(`/api/v1/inventory/${itemId}/movements`);
    return items.map(mapMovement);
  } catch {
    return [];
  }
}
